core/lib/zksync_core/src/state_keeper/io/common/mod.rs

```diff
@@ -0,0 +1,116 @@
+use std::time::Duration;
+
+use anyhow::Context;
+use multivm::interface::{L1BatchEnv, SystemEnv};
+use zksync_dal::StorageProcessor;
+use zksync_types::{L1BatchNumber, MiniblockNumber, H256};
+
+use super::PendingBatchData;
+
+#[cfg(test)]
+mod tests;
+
+/// Returns the amount of iterations `delay_interval` fits into `max_wait`, rounding up.
+pub(crate) fn poll_iters(delay_interval: Duration, max_wait: Duration) -> usize {
+    let max_wait_millis = max_wait.as_millis() as u64;
+    let delay_interval_millis = delay_interval.as_millis() as u64;
+    assert!(delay_interval_millis > 0, "delay interval must be positive");
+
+    ((max_wait_millis + delay_interval_millis - 1) / delay_interval_millis).max(1) as usize
+}
+
+/// Cursor of the miniblock / L1 batch progress used by [`StateKeeperIO`](super::StateKeeperIO) implementations.
+#[derive(Debug)]
+pub(crate) struct IoCursor {
```

Comment:
do I understand it right that now you use this struct both in EN and on the Main Node? Very nice that we are able to reuse!

Reply:
Yes, it's used in both I/O implementations.